### ISSUE TITLE
Ubuntu Xenial rsyslog package and logrotate

### ISSUE THREAD

**Issue by jeanpaulgalea** [rsyslog/rsyslog#1506]
Hi,

The rsyslog package for Ubuntu Xenial includes `/etc/logrotate.d/rsyslog` with a postrotate command of `reload rsyslog >/dev/null 2>&1 || true`.

Ubuntu Xenial comes with systemd by default and `reload` is only available when upstart is installed.

I think the postrotate command should be `invoke-rc.d rsyslog rotate >/dev/null` instead? As far as I understand, `invoke-rc.d` works with any init system.

In my opinion, `2>&1 || true` should be avoided, because in cases such as above, errors are not found unless by accident - e.g. rsyslog keeps writing to rotated files, rather than getting a cron email.

**Comment by rgerhards** [comment]
@friedl did you change this? I think I remember so. In that case, we can close this issue here. Please let us know.

**Comment by rgerhards** [comment]
Thanks for opening this issue. Unfortunately, I was hyper-busy with many other things for quite some month and not all issues were properly followed up. I am currently triaging everything.

@friedl if not already done, this would be good to have with the next release

**Comment by friedl** [comment]
Changed the logrotation script according to the suggestion for Xenial, Zesty, Artful. This will be available in the next package cycle.

https://github.com/rsyslog/rsyslog-pkg-ubuntu/commit/30d580842cf09e982317e608abac1268a64507d4

**Comment by deoren** [comment]
@friedl If you wish to also include Ubuntu 14.04 support within the same logrotate conf file, please see my remarks on rsyslog/rsyslog-pkg-ubuntu#56. Using the init.d script from that conf file appears to work equally well on the Ubuntu 14.04 and 16.04 LTS releases; I have been using that approach for many months now.

**Comment by friedl** [comment]
@deoren Thank you for the hint. I'll update the commit right away.

**Comment by deoren** [comment]
> @deoren Thank you for the hint. I'll update the commit right away.

Did you have a chance to make that change? I ask because I installed the latest rsyslog `rsyslog-8.31.0-0adiscon2trusty1` package from `ppa:adiscon/v8-stable` and found that the included `/etc/logrotate.d/rsyslog` conf file is broken for Ubuntu 14.04 systems.

## Error using provided /etc/logrotate.conf/rsyslog conf file

Here is the result of attempting to run the provided postrotate command on that test Ubuntu 14.04 box:

`root@ubuntu1404-virtual-machine:~# invoke-rc.d rsyslog rotate`
```
initctl: invalid command: rotate
Try `initctl --help' for more information.
invoke-rc.d: initscript rsyslog, action "rotate" failed.
```

## Stock Ubuntu 14.04 /etc/logrotate.d/rsyslog postrotate command

```
	postrotate
		reload rsyslog >/dev/null 2>&1 || true
	endscript

```

## Stock Ubuntu 16.04 /etc/logrotate.d/rsyslog postrotate command

```
	postrotate
		invoke-rc.d rsyslog rotate > /dev/null
	endscript

```

## Recommended postrotate command for Ubuntu 14.04+

```
    postrotate
        /etc/init.d/rsyslog rotate > /dev/null
    endscript
```

This results in the init.d script doing the heavy lifting; from everything I can tell this approach works equally well on Ubuntu 14.04 and Ubuntu 16.04 LTS releases.

This recommendation is based on the assumption that you have the requirement of using the same `/etc/logrotate.d/rsyslog` conf file for all supported versions of Ubuntu. If you don't, then the stock postrotate commands may be the way to go. 

Please see my notes on rsyslog/rsyslog-pkg-ubuntu#56 for additional information.

**Comment by friedl** [comment]
I did also add the same change to the 14.04 logrotate file, as per your suggestion.

The init script is actually separate for each platform. But, the actual content is the same. At least for 14.04 and 16.04. 

I updated the logrotate script per recommendation for 14.04 now. Please verify the new package.

**Comment by deoren** [comment]
> @friedl: I updated the logrotate script per recommendation for 14.04 now. Please verify the new package.

Confirmed for Ubuntu 14.04 (`rsyslog-8.31.0-0adiscon3trusty1`):

`root@ubuntu1404-virtual-machine:~# grep 'postrotate' -A 2 /etc/logrotate.d/rsyslog`
```
	postrotate
		/etc/init.d/rsyslog rotate > /dev/null	
	endscript
--
	postrotate
		/etc/init.d/rsyslog rotate > /dev/null	
	endscript
```

**Run command as-is**

`root@ubuntu1404-virtual-machine:~# /etc/init.d/rsyslog rotate > /dev/null`
`root@ubuntu1404-virtual-machine:~# echo $?`
```
0
```

**Run command, display output.**

`root@ubuntu1404-virtual-machine:~# /etc/init.d/rsyslog rotate`
```
 * Closing open files rsyslogd                         [ OK ]
```

**Comment by friedl** [comment]
Yes, this is solved.

**Comment by kitchen** [comment]
Curious: this doesn't appear to be in the latest rsyslog on 16.04?

I ask because I'm running into issues with `invoke-rc.d` identical to https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=795255

A chef cookbook I'm using is installing a /usr/sbin/policy-rc.d file to prevent a service from starting on package installation and the presence of that file seems to trigger that bug. It sounds like it's been fixed in 14.04, can that get fixed in 16.04 as well?

**Comment by kitchen** [comment]
Additionally, this same command to reload rsyslog appears in the ufw logrotate, at least on my 16.04 machine here.

**Comment by friedl** [comment]
The change was not carried over to 16.04, because the original request was to fix the logrotate in 14.04. And because 16.04 was working (obviously) fine, there was no need to go away from the system default.

The command seems to be executed fine either way, so I am not sure why we should change this in 16.04. Can you please elaborate this a little further? I'd rather not go away from the system default and potentially break something for everyone else.
